Progress Data Refactor Part 2

cforkish opened the pull request:
This PR builds on #37482, updating all the code that consumes the progress data to handle the new format. Note that there is still more work needed to get all the tests to pass.

This PR will be easier to review once #37482 is merged into `cforkish/progress-data-refactor` (not `staging`) since currently it contains all the commits from that PR as well. Once both PRs are merged into that base branch I will open a final PR to merge all the work into `staging`.

cforkish commented:
per Clare's suggestion, i changed the base for this PR to be Progress Data Refactor Part 1 so it should be easier to review, while still being possible to test the whole thing in action if desired.

jmkulwik commented on apps/src/templates/progress/ProgressBubbleSet.jsx:
```diff
@@ -87,16 +87,17 @@ class ProgressBubbleSet extends React.Component {
       hideAssessmentIcon
     } = this.props;
-
+    const levelProgress = studentProgress[level.id];
```
Here, it looks like studentProgress is an array, but above, it's an object. Was that intentional?

jmkulwik replied:
Copying an offline conversation here: 
PropTypes.objectOf means that all the values of the object conform to the type specified so that’s saying that studentProgress is an object with any type of key (in this case, level.id), and with values of type studentLevelProgressType

jmkulwik replied:
✔️

jmkulwik commented on apps/src/StudioApp.js:
```diff
@@ -3423,8 +3425,10 @@ StudioApp.prototype.isNotStartedLevel = function(config) {
-      progress.levelProgress[progress.currentLevelId] === undefined
+      (config.readonlyWorkspace &&
+        progress.progressByLevel[progress.currentLevelId] === undefined) ||
+      progress.progressByLevel[progress.currentLevelId].status ===
```
Noticing some test failures because `progress.progressByLevel[progress.currentLevelId]` is undefined (also because `progress` is undefined, but that's a testing issue since `progress` was shortcut by `config.readonlyWorkspace` and we never needed to mock it in the past). 
Given that `progress.progressByLevel[progress.currentLevelId]` is expected to sometimes be undefined in 3429, should we also be checking if it's undefined in 3430?

jmkulwik replied:
I realize that comment got a bit self-referential. The short story: I think 3430 should be  `progress.progressByLevel[progress.currentLevelId] &&  progress.progressByLevel[progress.currentLevelId].status` or something similar.

jmkulwik replied:
✔️

jmkulwik commented on apps/src/code-studio/components/progress/LessonProgress.jsx:
```diff
@@ -160,16 +166,26 @@ class LessonProgress extends Component {
-    levels = levels.filter(level => !level.bonus);
+    const location = window.location.href;
+    const currentPage = parseInt(
+      location.substring(location.lastIndexOf('/') + 1)
```
I think you'll also need to set an end point. urls can also have `?section=###` or `?enableExperiments=name`

On that note, this feels fragile. How do we set this number? Can we leverage that method to get this int?

We don't necessarily have to fix this to ship it, but a linked follow-up task would be nice.

cforkish replied:
Here is a ticket to track this tech debt: https://codedotorg.atlassian.net/browse/LP-1641

From MDN: "If parseInt encounters a character that is not a numeral in the specified radix, it ignores it and all succeeding characters and returns the integer value parsed up to that point."

So luckily in this case this is not quite as fragile as it looks. Without a significant change to how we're doing things, the page number will always be the last path component of this url, and anything after it will be ignored. That said, I agree this is hacky, but it was the simplest solution I could think of.

jmkulwik replied:
Oh cool! Thanks for the explanation - didn't realize that was how parseInt worked.

jmkulwik commented on apps/src/code-studio/components/progress/TeacherPanel.jsx:
```diff
@@ -76,10 +82,69 @@ class TeacherPanel extends React.Component {
+  /**
+   * This is the only place in our code where we get levels from
+   * ScriptLevel.summarize_for_teacher_panel or
+   * ScriptLevel.summarize_as_bonus_for_teacher_panel,
```
Is this comment still relevant? (asking since summarize_for_teacher_panel etc isn't in this function)

cforkish replied:
haha, this comment is brand new! those summarize methods are in the rails code, which i thought would be clear from the names and mentioning the rails code at the end of this comment, but i can flesh this out with a little more explanation.

jmkulwik replied:
Ahhh ok. I think I didn't read the whole comment.

jmkulwik commented on apps/src/templates/progress/ProgressBubble.jsx:
```diff
@@ -88,17 +92,11 @@ const styles = {
-      PropTypes.number
-    ]),
+    selectedSectionId: PropTypes.number,
+    selectedStudentId: PropTypes.number,
```
🎉

jmkulwik commented on apps/src/templates/progress/progressTestHelpers.js:
```diff
@@ -44,6 +46,17 @@ export const fakeLevels = (numLevels, {startLevel = 1, named = true} = {}) =>
     return fakeLevel(overrideData);
   });
 
+export const fakeProgressForLevels = (
```
This has made writing tests _so_ much easier

cforkish commented:
i want to call attention to [this commit](https://github.com/code-dot-org/code-dot-org/pull/37483/commits/60e26ac83062e3cb1275c8178fd20809f83b85f2) -- i found a couple cases where the simplification i had made of not looking at per-page progress for multi-page assessments was insufficient.

background: i had to standardize our level ids to be numbers for consistent key comparison, but there were a few cases where we relied on being able to use strings for level ids. in the case of multi-page assessments, we were using levelId_pageNumber for per-page ids. initially, i found that we didn’t need that extra layer of granularity because we don’t track per-page assessment results, so the progress status was always the same for all pages. however, it turns out that we do keep track of the not_tried/attempted status on a per-page level, and that turns out to be useful (or at least, needs to be handled), but we can no longer use per-page level ids since we can’t encode them in a number without risking collisions with other level ids.

1. when working through an assessment, we need to know which page we’re on to highlight it it in the header. for this i added a slight hack of getting the page number from the url and comparing that to the level number. see my comment in LessonProgress.jsx line 217.

2. when a student starts working on an assessment, that level gets marked with the status attempted, so all the bubbles for all the pages were getting a green border. i noticed that the progress object we get from the server for multi-page assessments includes a pages_completed array which contains a result value for each page (or null if not attempted yet), so i added a recursive pages property to studentLevelProgressType to inflate those result values into full studentLevelProgressType objects, so now for assessment levels we use the per-page progress by getting studentLevelProgress.pages[level.levelNumber-1]

jmkulwik commented on apps/src/code-studio/progress.js:
```diff
@@ -120,11 +120,7 @@ progress.generateStageProgress = function(
   store.dispatch(
-    mergeProgressResults(
-      _.mapValues(progressData.levels, level =>
-        level.submitted ? TestResults.SUBMITTED_RESULT : level.result
```
Does level.submitted get used anywhere anymore? (I'm seeing it in tests and I'm not sure whether I can get rid of it)

jmkulwik commented on apps/src/templates/progress/progressTypes.js:
```diff
@@ -50,13 +50,23 @@ export const levelType = PropTypes.shape({
  * A boolean indicating if a student was paired on a level.
```
In the `levelType` definition above, can you add in some of the other props that are sometimes in levels?
driver, navigator, contained, submitLevel, submitted, assessment, passed, user_id

cforkish replied:
sure. similar to your question about level.submitted above, i'll have to audit our use of all these properties though. i think we mostly rely on `progress.submitted` and `level.levelKind === LevelKind.assessment` instead of `level.submitted` and `level.assessment` for example.

jmkulwik replied:
Oh interesting. Would it make sense to deprecate the others then? i.e. `level.submitted` -> `level.__deprecatedSubmitted`
Maybe as a follow-up task.

jmkulwik commented on apps/src/code-studio/components/progress/LessonProgress.jsx:
```diff
@@ -194,24 +210,44 @@ class LessonProgress extends Component {
-                />
-              </div>
-            ))}
+            {filteredLevels.map((level, index) => {
```
@cforkish Do you by chance have an example level that hits this code path? I'm hitting something odd in storybook where map doesn't seem to be working as expected and I'd like to compare it to a non-storybook example.

jmkulwik replied:
Turns out this code path gets hit in lots of places.
Here's one: http://studio.code.org/s/coursee-2019/stage/7/puzzle/4
